common/events_io.hpp

```diff
@@ -126,18 +126,21 @@ class AsyncWriterFromWriter : virtual public mio::AsyncWriter {
 	EventLoop &loop_;
 };
 
-using AsyncReaderFromEventLoopFunc = function<mio::AsyncReaderPtr(EventLoop &loop)>;
+using AsyncReaderFromEventLoopFunc = function<mio::ExpectedAsyncReaderPtr(EventLoop &loop)>;
 
 class ReaderFromAsyncReader : virtual public mio::Reader {
 public:
 	// Since the async reader usually needs the event loop object, we need to pass in the
 	// AsyncReaderFromEventLoopFunc factory function to create the async reader inside of the
-	// constructor.
-	ReaderFromAsyncReader(AsyncReaderFromEventLoopFunc func);
+	// constructor. Because this can sometimes fail, use a factory function which can return
+	// errors, so that we don't have to use exceptions.
+	static mio::ExpectedReaderPtr Construct(AsyncReaderFromEventLoopFunc func);
```

Comment:
Now it actually, finally, makes sense to me! With what I imagine being *a factory function*.